### DISCUSSION TITLE
Fox ESS RS485 Battery to BMS

### DISCUSSION THREAD

**Discussion by webbey01** [StealthChesnut/HA-FoxESS-Modbus#50] [Q&A]
Hi there

I would like to read the information between the battery and the BMS unit hopefully make up a little widget that tells me what the state of the actual battery is in, has anyone had any luck reading this information and decoding it?

**Comment by rand12345** [comment]
I have mapped the can bus data frames for a V1 system, the details are in [this repo.](https://github.com/rand12345/solax_can_bus)


@StealthChesnut Would you like to collaborate on an extended modbus project? Thinking about mapping all cell level data to a custom set of registers which can then be integrated into HA, existing FoxESS registers can pass through. This should also solve some ethernet issues users are having.

Something like this is doable on an ESP32C3, there's enough IO and the entire setup can be powered from the inverter 3.3v supply on Com.


```mermaid
graph LR
A[Master BMS] -- Can BMS --> D
 B((ESP32 Gateway)) -- Can Com --> D(FoxESS)
B -- Modbus RTU --> D
E[WiFi] -- Modbus TCP --> B
```

Alternatively, you could set it up like this with dual can and intercept charge/discharge rates. Pretty useful for customising battery dynamics.

```mermaid
graph LR
A[Master BMS] -- Can BMS--> B((ESP32 Gateway))
B -- Can Com --> D(FoxESS)
B -- Modbus RTU --> D
E[WiFi] -- Modbus TCP --> B
```

**Reply by StealthChesnut** [reply]
This looks very interesting @rand12345 - would like to discuss with you. Are you on the foxess community forums? Any chance you could ping me a message on there please?

**Reply by rand12345** [reply]
Will PM you shortly.

**Reply by rand12345** [reply]
https://foxesscommunity.com/viewtopic.php?t=145
